File: app/src/main/java/com/example/review/DataStructureFile/TreeNode.java
package com.example.review.DataStructureFile;

import java.util.ArrayList;
import java.util.Objects;

public class TreeNode<TYPE_VALUE, TYPE_INDEX> {
    public TreeNode() {

    }

    public TreeNode(TYPE_VALUE value) {
        this.value = value;
    }

    public TYPE_VALUE value;

    public TYPE_INDEX index;

    public ArrayList<TreeNode<TYPE_VALUE, TYPE_INDEX>> node;

    @Override
    public String toString() {
        return value.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        TreeNode<?, ?> treeNode = (TreeNode<?, ?>) o;
        return Objects.equals(value, treeNode.value);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value);
    }
}

File: app/src/main/java/com/example/review/DataStructureFile/TreeNode.kt
package com.example.review.DataStructureFile

import java.util.*

class TreeNode<TYPE_VALUE, TYPE_INDEX> {
    constructor() {}
    constructor(value: TYPE_VALUE) {
        this.value = value
    }

    var value: TYPE_VALUE? = null
    var index: TYPE_INDEX? = null
    var node: ArrayList<TreeNode<TYPE_VALUE, TYPE_INDEX>>? = null
    override fun toString(): String {
        return value.toString()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) return true
        if (o == null || javaClass != o.javaClass) return false
        val treeNode = o as TreeNode<*, *>
        return value == treeNode.value
    }

    override fun hashCode(): Int {
        return Objects.hash(value)
    }
}